src/main/scala/org/ergoplatform/nodeView/state/UtxoStateReader.scala

```diff
@@ -38,22 +39,23 @@ trait UtxoStateReader extends ErgoStateReader with TransactionValidation {
     */
   def validateWithCost(tx: ErgoTransaction,
                        stateContextOpt: Option[ErgoStateContext],
-                       complexityLimit: Int,
+                       costLimit: Long,
                        interpreterOpt: Option[ErgoInterpreter]): Try[Long] = {
     val context = stateContextOpt.getOrElse(stateContext)
 
     val verifier = interpreterOpt.getOrElse(ErgoInterpreter(context.currentParameters))
 
     tx.statelessValidity().flatMap { _ =>
       val boxesToSpend = tx.inputs.flatMap(i => boxById(i.boxId))
-      val txComplexity = boxesToSpend.map(_.ergoTree.complexity).sum
-      if (txComplexity > complexityLimit) {
-        throw new Exception(s"Transaction $tx has too high complexity $txComplexity")
-      }
       tx.statefulValidity(
-        boxesToSpend,
-        tx.dataInputs.flatMap(i => boxById(i.boxId)),
-        context)(verifier)
+          boxesToSpend,
+          tx.dataInputs.flatMap(i => boxById(i.boxId)),
+          context)(verifier).flatMap {
+            case txCost if txCost > costLimit =>
```

Comment:
At the time we check this limit, the node may already have spent full MaxBlockCost of computations. If the ratio between MaxBlockCost and MaxTxCost become higher, then this is increasingly may be an attack vector. See my previous comment.

Reply:
good catch, made a new version of *validateWithCost* which is passing accumulated cost considering the limit and tests, please check

@aslesarenko made 
```case Failure(mme: MalformedModifierError) if mme.message.contains("CostLimitException")``` as no structured information to extract CostLimitException available